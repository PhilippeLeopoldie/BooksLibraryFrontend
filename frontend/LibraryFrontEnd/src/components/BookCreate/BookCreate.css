.book-create-form__container {
    margin: 2rem auto auto auto;
    background-color: inherit;
    display: flex;
    justify-content: center;
    max-width: 20rem;
    max-height: 29rem;
    height: 362px;
    padding: 1rem;
    position: relative;
    border-radius: 10px;
    border: 1px var(--forminputcolor) solid;
}

.input--black, .select--black, .option--black {
  color: #ffffff;
}

.input, .select {
    background-color: inherit;
    width: 90%;
    border: 1px var(--forminputcolor) solid;
}

.option--black {
    background-color: var(--backgroundcolor-darkmode);
}

.select {
    position: relative;
}

.book-create-form__container--natural {
    border: 1px var(--cardbordercolor_light) solid;
    & .input {
        border: 1px var(--cardbordercolor_light) solid;
    }
}



.input__imageUrl--black,
.input__imageUrl--natural {
    font-size: large;
    text-decoration: underline;
    color: #4ba5ff;
}

.footer--flex {
  display: flex;
  justify-content: space-around;
}


.book-create-form {
  background-color: transparent;
  display: grid;
  justify-content: initial;
  grid-template-rows: repeat(5, 1fr);
  width: 90%;
}

.validation__error-message {
  color: red;
  background-color: transparent;
  font-size: 1.2rem;
  font-weight: bold;
}

.book-create-form__output-card {
  background-color: transparent;
}

.book-create-form__output {
  background-color: transparent;
  color: green;
  font-size: 2rem;
  font-weight: bold;
}
.book-create-form__output--back-to-home--black {
  background-color: transparent;
  margin: 0;
  color: white;
  text-decoration: underline;
}

.book-create-form__output--back-to-home--natural {
  background-color: transparent;
  margin: 0;
}

.footer--flex {
  display: flex;
  justify-content: space-around;
  background-color: transparent;
}

.book-create-form__cancel-button {
    font-size: 1.2rem;
    margin-top: 1rem;
    margin-bottom: 1.5rem;    
}

.book-create-form__post-button {
  font-weight: bold;
  font-size: 1.2rem;
  margin-top: 1rem;
  margin-bottom: 1.5rem;  
}

