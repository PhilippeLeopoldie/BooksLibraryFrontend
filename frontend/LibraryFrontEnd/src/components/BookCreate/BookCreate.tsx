import { BookType } from "../../constants/types";
import { useContext, useState } from "react";
import "./BookCreate.css";
import { Link } from "react-router-dom";
import { BOOK_URL } from "../../constants/api";
import { genresCacheContext, paginatedBooksCacheContext, topBooksCacheContext, ThemeContext } from "../../App/App";

export const BookCreate = () => {
    const theme = useContext(ThemeContext);
    const paginatedBooksCache = useContext(paginatedBooksCacheContext);
    const topBooksCache = useContext(topBooksCacheContext);
    const genresCache = useContext(genresCacheContext);
    const genresListCache = genresCache?.genresCache?.genres;
    const [formData, setFormData] = useState<BookType>({
        id:0,
        title: "",
        author: "",
        description: "",
        imageUrl: "",
        averageRate: 0,
        genreId: ""
    });

    const [bookCreatedMessage, setBookCreatedMessage] = useState<boolean>(false);
    const [errorBook, setErrorBook] = useState<Boolean>(false);
    const [errorBookDetail, setErrorBookDetail] = useState<string>("");
    const rootElement = document.documentElement;

    const PostBook = async () => {
        const requestOptions = {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify({
                title: formData.title,
                author: formData.author,
                imageUrl: formData.imageUrl,
                genreId: formData.genreId
            }),
        };
        const bookResponse: Response = await fetch(
            BOOK_URL,
            requestOptions
        );
        if (bookResponse.status === 201) {
            setErrorBook(false);
            const body = await bookResponse.json();
            setBookCreatedMessage(true);
            paginatedBooksCache?.setPaginatedBooks({
                paginatedItems: [],
                totalItems: 0,
                page: 0,
                totalPages: 0,
                requestedAt: ""
            });
            topBooksCache?.setTopBooksCache([]);
            setFormData({
                id: 0,
                title: "",
                author: "",
                description: "",
                imageUrl: "",
                averageRate: 0,
                genreId:""
            });
            return body;
        } else if (bookResponse.status === 400) {
            setErrorBook(true);
            const errorData = await bookResponse.json();
            setErrorBookDetail(errorData.detail);
        }
    };
    const HideBookCreatedMessage = () => {
        setBookCreatedMessage(false);
    };

    const HideErrorDetail = () => {
        setErrorBook(false);
    };

    const HandleInputChange = (
        e:
            | React.ChangeEvent<HTMLInputElement>
            | React.ChangeEvent<HTMLTextAreaElement>
            | React.ChangeEvent<HTMLSelectElement>
    ) => {
        setFormData({ ...formData, [e.target.name]: e.target.value });
        bookCreatedMessage && HideBookCreatedMessage();
        errorBook && HideErrorDetail();
    };

    return (
        <>
            <section
                className={`book-create-form__container--${theme} book-create-form__container`}
            >
                <form
                    onSubmit={(e) => e.preventDefault()}
                    action={BOOK_URL}
                    method="POST"
                    className="book-create-form"
                >
                    <input
                        className={`input input--${theme}`}
                        placeholder="Title"
                        name="title"
                        value={formData.title}
                        onChange={(e) => HandleInputChange(e)}
                    />

                    <input
                        className={`input input--${theme}`}
                        placeholder="Author"
                        name="author"
                        value={formData.author}
                        onChange={(e) => HandleInputChange(e)}
                    />
                    
                    <select
                        className={`select select--${theme}` }
                        name="genreId"
                        value={formData.genreId}
                        onChange={(e) => HandleInputChange(e)}
                    >
                        <option className={`option option--${theme}`} value="" disabled>Choose a genre</option>
                        {genresListCache && genresListCache.map((genre) => (
                            <option className={`option option--${theme}`} key={genre.id} value={genre.id}> {genre.name} </option>
                        ))
                        }
                    </select>

                    <input
                        className={`input input--${theme} input__imageUrl--${theme}`}
                        placeholder="Paste Book Image Link"
                        name="imageUrl"
                        value={formData.imageUrl}
                        onChange={(e) => HandleInputChange(e)}
                    />
                    <section className="book-create-form__output-card">
                        {errorBook && (
                            <div className="validation__error-message">{errorBookDetail}</div>
                        )}
                        {bookCreatedMessage && (
                            <div className="book-create-form__output-card">
                                <div className="book-create-form__output">Book created!</div>
                                <Link to="/">
                                    <h3 className={`book-create-form__output--back-to-home--${theme}`}>
                                        back to Home
                                    </h3>
                                </Link>
                            </div>
                        )}
                    </section>
                    <footer className="footer--flex">
                        <button
                            className="button book-create-form__cancel-button"
                            onClick={() => {
                                setErrorBook(false);
                                setFormData({
                                    id: 0,
                                    title: "",
                                    author: "",
                                    description: "",
                                    imageUrl: "",
                                    averageRate: 0,
                                    genreId: ""
                                });
                            }}
                        >
                            Cancel
                        </button>
                        <button
                            className="button book-create-form__post-button"
                            onClick={async () => {
                                await PostBook();
                            }}
                        >
                            Add Book
                        </button>
                    </footer>
                </form>
            </section>
        </>
    );
};
