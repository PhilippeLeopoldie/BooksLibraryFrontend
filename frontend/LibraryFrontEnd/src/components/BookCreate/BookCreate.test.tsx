import {render, screen,} from "@testing-library/react";
import { BookCreate } from "./BookCreate";
import "@testing-library/jest-dom";

describe("Book to Create", () => {
  test("snapShot: render Book Create", ()=> {
    const {asFragment} = render(<BookCreate/>)

    expect(asFragment()).toMatchSnapshot();
  })
})