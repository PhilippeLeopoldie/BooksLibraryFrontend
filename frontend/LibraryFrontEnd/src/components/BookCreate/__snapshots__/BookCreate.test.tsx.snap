// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book to Create snapShot: render Book Create 1`] = `
<DocumentFragment>
  <div
    class="book-create-form__container--natural book-create-form__container"
  >
    <script>
      #f3f3f4
    </script>
    <form
      action="https://booklibrary-backend-20f7a19cecb2.herokuapp.com/api/Books"
      class="book-create-form"
      method="POST"
    >
      <input
        class="input input--natural"
        name="title"
        placeholder="Title"
        value=""
      />
      <input
        class="input input--natural"
        name="author"
        placeholder="Author"
        value=""
      />
      <input
        class="input input--natural input__imageUrl--natural"
        name="imageUrl"
        placeholder="Book Image link"
        value=""
      />
      <div
        class="book-create-form__output-card"
      />
      <footer
        class="footer--flex"
      >
        <button
          class="button book-create-form__cancel-button"
        >
          Cancel
        </button>
        <button
          class="button book-create-form__post-button"
        >
          Create book
        </button>
      </footer>
    </form>
  </div>
</DocumentFragment>
`;
